File: sdks/thunkmetrc-py/wrapper/src/thunkmetrc/wrapper/services/items_service.py
from typing import Any, Optional, List, AsyncGenerator
from thunkmetrc.client import MetrcClient
from ..ratelimiter import MetrcRateLimiter
from ..models import *
from ..utils import iterate_pages, deserialize_page

class ItemsService:
    def __init__(self, client: MetrcClient, limiter: MetrcRateLimiter):
        self.client = client
        self.limiter = limiter

    async def create_items_brand(self, body: Optional[List['ItemsCreateBrandRequest']] = None, license_number: Optional[str] = None) -> 'WriteResponse':
        """
        Creates one or more new item brands for the specified Facility identified by the License Number.
        Permissions Required:
        - Manage Items
        POST /items/v2/brand
        Parameters:
        - body (List['ItemsCreateBrandRequest']): Request body
        - license_number (str, optional): Filter by licenseNumber
        """
        async def op():
            return await self.client.items.create_items_brand(body=body,
                license_number=license_number,
            )
        return await self.limiter.execute(None, False, op)

    async def create_items_file(self, body: Optional[List['ItemsCreateFileRequest']] = None, license_number: Optional[str] = None) -> 'WriteResponse':
        """
        Uploads one or more image or PDF files for products, labels, packaging, or documents at the specified Facility.
        Permissions Required:
        - Manage Items
        POST /items/v2/file
        Parameters:
        - body (List['ItemsCreateFileRequest']): Request body
        - license_number (str, optional): Filter by licenseNumber
        """
        async def op():
            return await self.client.items.create_items_file(body=body,
                license_number=license_number,
            )
        return await self.limiter.execute(None, False, op)

    async def create_items(self, body: Optional[List['ItemsCreateItemsRequest']] = None, license_number: Optional[str] = None) -> 'WriteResponse':
        """
        Permissions Required:
        - Manage Items
        POST /items/v2
        Parameters:
        - body (List['ItemsCreateItemsRequest']): Request body
        - license_number (str, optional): Filter by licenseNumber
        """
        async def op():
            return await self.client.items.create_items(body=body,
                license_number=license_number,
            )
        return await self.limiter.execute(None, False, op)

    async def create_items_photo(self, body: Optional[List['ItemsCreatePhotoRequest']] = None, license_number: Optional[str] = None) -> 'WriteResponse':
        """
        This endpoint allows only BMP, GIF, JPG, and PNG files and uploaded files can be no more than 5 MB in size.
        Permissions Required:
        - Manage Items
        POST /items/v2/photo
        Parameters:
        - body (List['ItemsCreatePhotoRequest']): Request body
        - license_number (str, optional): Filter by licenseNumber
        """
        async def op():
            return await self.client.items.create_items_photo(body=body,
                license_number=license_number,
            )
        return await self.limiter.execute(None, False, op)

    async def delete_items_brand_by_id(self, id: str, body: Any = None, license_number: Optional[str] = None) -> Any:
        """
        Archives the specified Item Brand by Id for the given Facility License Number.
        Permissions Required:
        - Manage Items
        DELETE /items/v2/brand/{id}
        Parameters:
        - id (str): Path parameter id
        - license_number (str, optional): Filter by licenseNumber
        """
        async def op():
            return await self.client.items.delete_items_brand_by_id(id, body=body,
                license_number=license_number,
            )
        return await self.limiter.execute(None, False, op)

    async def delete_items_by_id(self, id: str, body: Any = None, license_number: Optional[str] = None) -> Any:
        """
        Archives the specified Product by Id for the given Facility License Number.
        Permissions Required:
        - Manage Items
        DELETE /items/v2/{id}
        Parameters:
        - id (str): Path parameter id
        - license_number (str, optional): Filter by licenseNumber
        """
        async def op():
            return await self.client.items.delete_items_by_id(id, body=body,
                license_number=license_number,
            )
        return await self.limiter.execute(None, False, op)

    async def get_active_items(self, body: Any = None, last_modified_end: Optional[str] = None, last_modified_start: Optional[str] = None, license_number: Optional[str] = None, page_number: Optional[str] = None, page_size: Optional[str] = None) -> PaginatedResponse['Item']:
        """
        Returns a list of active items for the specified Facility.
        Permissions Required:
        - Manage Items
        GET /items/v2/active
        Parameters:
        - last_modified_end (str, optional): Filter by lastModifiedEnd
        - last_modified_start (str, optional): Filter by lastModifiedStart
        - license_number (str, optional): Filter by licenseNumber
        - page_number (str, optional): Filter by pageNumber
        - page_size (str, optional): Filter by pageSize
        """
        async def op():
            return await self.client.items.get_active_items(body=body,
                last_modified_end=last_modified_end,
                last_modified_start=last_modified_start,
                license_number=license_number,
                page_number=page_number,
                page_size=page_size,
            )
        return await self.limiter.execute(None, True, op)

    async def get_items_brands(self, body: Any = None, license_number: Optional[str] = None, page_number: Optional[str] = None, page_size: Optional[str] = None) -> PaginatedResponse['Brand']:
        """
        Retrieves a list of active item brands for the specified Facility.
        Permissions Required:
        - Manage Items
        GET /items/v2/brands
        Parameters:
        - license_number (str, optional): Filter by licenseNumber
        - page_number (str, optional): Filter by pageNumber
        - page_size (str, optional): Filter by pageSize
        """
        async def op():
            return await self.client.items.get_items_brands(body=body,
                license_number=license_number,
                page_number=page_number,
                page_size=page_size,
            )
        return await self.limiter.execute(None, True, op)

    async def get_items_categories(self, body: Any = None, license_number: Optional[str] = None, page_number: Optional[str] = None, page_size: Optional[str] = None) -> PaginatedResponse['Category']:
        """
        Retrieves a list of item categories.
        GET /items/v2/categories
        Parameters:
        - license_number (str, optional): Filter by licenseNumber
        - page_number (str, optional): Filter by pageNumber
        - page_size (str, optional): Filter by pageSize
        """
        async def op():
            return await self.client.items.get_items_categories(body=body,
                license_number=license_number,
                page_number=page_number,
                page_size=page_size,
            )
        return await self.limiter.execute(None, True, op)

    async def get_items_file_by_id(self, id: str, body: Any = None, license_number: Optional[str] = None) -> 'File':
        """
        Retrieves a file by its Id for the specified Facility.
        Permissions Required:
        - Manage Items
        GET /items/v2/file/{id}
        Parameters:
        - id (str): Path parameter id
        - license_number (str, optional): Filter by licenseNumber
        """
        async def op():
            return await self.client.items.get_items_file_by_id(id, body=body,
                license_number=license_number,
            )
        return await self.limiter.execute(None, True, op)

    async def get_inactive_items(self, body: Any = None, license_number: Optional[str] = None, page_number: Optional[str] = None, page_size: Optional[str] = None) -> PaginatedResponse['Item']:
        """
        Retrieves a list of inactive items for the specified Facility.
        Permissions Required:
        - Manage Items
        GET /items/v2/inactive
        Parameters:
        - license_number (str, optional): Filter by licenseNumber
        - page_number (str, optional): Filter by pageNumber
        - page_size (str, optional): Filter by pageSize
        """
        async def op():
            return await self.client.items.get_inactive_items(body=body,
                license_number=license_number,
                page_number=page_number,
                page_size=page_size,
            )
        return await self.limiter.execute(None, True, op)

    async def get_items_by_id(self, id: str, body: Any = None, license_number: Optional[str] = None) -> 'Item':
        """
        Retrieves detailed information about a specific Item by Id.
        Permissions Required:
        - Manage Items
        GET /items/v2/{id}
        Parameters:
        - id (str): Path parameter id
        - license_number (str, optional): Filter by licenseNumber
        """
        async def op():
            return await self.client.items.get_items_by_id(id, body=body,
                license_number=license_number,
            )
        return await self.limiter.execute(None, True, op)

    async def get_items_photo_by_id(self, id: str, body: Any = None, license_number: Optional[str] = None) -> 'Photo':
        """
        Retrieves an image by its Id for the specified Facility.
        Permissions Required:
        - Manage Items
        GET /items/v2/photo/{id}
        Parameters:
        - id (str): Path parameter id
        - license_number (str, optional): Filter by licenseNumber
        """
        async def op():
            return await self.client.items.get_items_photo_by_id(id, body=body,
                license_number=license_number,
            )
        return await self.limiter.execute(None, True, op)

    async def update_items_brand(self, body: Optional[List['ItemsUpdateBrandRequest']] = None, license_number: Optional[str] = None) -> 'WriteResponse':
        """
        Updates one or more existing item brands for the specified Facility.
        Permissions Required:
        - Manage Items
        PUT /items/v2/brand
        Parameters:
        - body (List['ItemsUpdateBrandRequest']): Request body
        - license_number (str, optional): Filter by licenseNumber
        """
        async def op():
            return await self.client.items.update_items_brand(body=body,
                license_number=license_number,
            )
        return await self.limiter.execute(None, False, op)

    async def update_items(self, body: Optional[List['ItemsUpdateItemsRequest']] = None, license_number: Optional[str] = None) -> 'WriteResponse':
        """
        Updates one or more existing products for the specified Facility.
        Permissions Required:
        - Manage Items
        PUT /items/v2
        Parameters:
        - body (List['ItemsUpdateItemsRequest']): Request body
        - license_number (str, optional): Filter by licenseNumber
        """
        async def op():
            return await self.client.items.update_items(body=body,
                license_number=license_number,
            )
        return await self.limiter.execute(None, False, op)

File: sdks/thunkmetrc-cs/ThunkMetrc.Wrapper/Services/ItemsService.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using ThunkMetrc.Client;
using Microsoft.Extensions.Logging;
using System.Text.Json;
using System.Threading;
using System.Runtime.CompilerServices;
using ThunkMetrc.Wrapper;
using ThunkMetrc.Wrapper.Models;
using File = ThunkMetrc.Wrapper.Models.File;

namespace ThunkMetrc.Wrapper.Services
{
    public class ItemsService(MetrcClient client, IMetrcRateLimiter rateLimiter)
    {
        private readonly MetrcClient _client = client;
        private readonly IMetrcRateLimiter _rateLimiter = rateLimiter;
        /// <summary>
        /// Creates one or more new item brands for the specified Facility identified by the License Number.
        /// Permissions Required:
        /// - Manage Items
        /// </summary>
        /// <param name="body">Request body.</param>
        /// <param name="licenseNumber">The License Number of the Facility for the Item Brands to create.</param>
        /// <param name="cancellationToken">The cancellation token to cancel the operation.</param>
        public async Task CreateItemsBrand(List<ItemsCreateBrandRequest>? body = null, string? licenseNumber = null, CancellationToken cancellationToken = default)
        {
            await _rateLimiter.ExecuteAsync<object?>(null, false, async () => { await _client.CreateItemsBrand(body, licenseNumber); return null; }, cancellationToken);
        }
        /// <summary>
        /// Uploads one or more image or PDF files for products, labels, packaging, or documents at the specified Facility.
        /// Permissions Required:
        /// - Manage Items
        /// </summary>
        /// <param name="body">Request body.</param>
        /// <param name="licenseNumber">Optional query parameter.</param>
        /// <param name="cancellationToken">The cancellation token to cancel the operation.</param>
        public async Task CreateItemsFile(List<ItemsCreateFileRequest>? body = null, string? licenseNumber = null, CancellationToken cancellationToken = default)
        {
            await _rateLimiter.ExecuteAsync<object?>(null, false, async () => { await _client.CreateItemsFile(body, licenseNumber); return null; }, cancellationToken);
        }
        /// <summary>
        /// Permissions Required:
        /// - Manage Items
        /// </summary>
        /// <param name="body">Request body.</param>
        /// <param name="licenseNumber">The License Number of the Facility for which to create new items.</param>
        /// <param name="cancellationToken">The cancellation token to cancel the operation.</param>
        public async Task CreateItems(List<ItemsCreateItemsRequest>? body = null, string? licenseNumber = null, CancellationToken cancellationToken = default)
        {
            await _rateLimiter.ExecuteAsync<object?>(null, false, async () => { await _client.CreateItems(body, licenseNumber); return null; }, cancellationToken);
        }
        /// <summary>
        /// This endpoint allows only BMP, GIF, JPG, and PNG files and uploaded files can be no more than 5 MB in size.
        /// Permissions Required:
        /// - Manage Items
        /// </summary>
        /// <param name="body">Request body.</param>
        /// <param name="licenseNumber">The License Number of the Facility for which to add an image.</param>
        /// <param name="cancellationToken">The cancellation token to cancel the operation.</param>
        public async Task CreateItemsPhoto(List<ItemsCreatePhotoRequest>? body = null, string? licenseNumber = null, CancellationToken cancellationToken = default)
        {
            await _rateLimiter.ExecuteAsync<object?>(null, false, async () => { await _client.CreateItemsPhoto(body, licenseNumber); return null; }, cancellationToken);
        }
        /// <summary>
        /// Archives the specified Item Brand by Id for the given Facility License Number.
        /// Permissions Required:
        /// - Manage Items
        /// </summary>
        /// <param name="id">id</param>
        /// <param name="licenseNumber">The License Number of the Facility for the Item Brand to delete.</param>
        /// <param name="cancellationToken">The cancellation token to cancel the operation.</param>
        public async Task DeleteItemsBrandById(string id, string? licenseNumber = null, CancellationToken cancellationToken = default)
        {
            await _rateLimiter.ExecuteAsync<object?>(null, false, async () => { await _client.DeleteItemsBrandById(id, licenseNumber); return null; }, cancellationToken);
        }
        /// <summary>
        /// Archives the specified Product by Id for the given Facility License Number.
        /// Permissions Required:
        /// - Manage Items
        /// </summary>
        /// <param name="id">id</param>
        /// <param name="licenseNumber">The License Number of the Facility for the Item to delete.</param>
        /// <param name="cancellationToken">The cancellation token to cancel the operation.</param>
        public async Task DeleteItemsById(string id, string? licenseNumber = null, CancellationToken cancellationToken = default)
        {
            await _rateLimiter.ExecuteAsync<object?>(null, false, async () => { await _client.DeleteItemsById(id, licenseNumber); return null; }, cancellationToken);
        }
        /// <summary>
        /// Returns a list of active items for the specified Facility.
        /// Permissions Required:
        /// - Manage Items
        /// </summary>
        /// <param name="lastModifiedEnd">The last modified end timestamp</param>
        /// <param name="lastModifiedStart">The last modified start timestamp</param>
        /// <param name="licenseNumber">The License Number of the Facility for which to return the list of active items.</param>
        /// <param name="pageNumber">The number of the data page from which to return data.</param>
        /// <param name="pageSize">The number of records to return per page. Pagination is currently disabled by default. You can enable pagination on this query by specifying a value that does not exceed 20.</param>
        /// <param name="cancellationToken">The cancellation token to cancel the operation.</param>
        public async Task<PaginatedResponse<Item>> GetActiveItems(string? lastModifiedEnd = null, string? lastModifiedStart = null, string? licenseNumber = null, string? pageNumber = null, string? pageSize = null, CancellationToken cancellationToken = default)
        {
            var result = await _rateLimiter.ExecuteAsync<object>(null, true, () => _client.GetActiveItems(lastModifiedEnd, lastModifiedStart, licenseNumber, pageNumber, pageSize), cancellationToken);
            if (result is JsonElement json)
            {
                return json.Deserialize<PaginatedResponse<Item>>() ?? throw new InvalidOperationException("Failed to deserialize response");
            }
            throw new InvalidOperationException("Unexpected response type");
        }
        /// <summary>
        /// Retrieves a list of active item brands for the specified Facility.
        /// Permissions Required:
        /// - Manage Items
        /// </summary>
        /// <param name="licenseNumber">The License Number of the Facility for which to return the list of active item brands.</param>
        /// <param name="pageNumber">The number of the data page from which to return data.</param>
        /// <param name="pageSize">The number of records to return per page. Pagination is currently disabled by default. You can enable pagination on this query by specifying a value that does not exceed 20.</param>
        /// <param name="cancellationToken">The cancellation token to cancel the operation.</param>
        public async Task<PaginatedResponse<Brand>> GetItemsBrands(string? licenseNumber = null, string? pageNumber = null, string? pageSize = null, CancellationToken cancellationToken = default)
        {
            var result = await _rateLimiter.ExecuteAsync<object>(null, true, () => _client.GetItemsBrands(licenseNumber, pageNumber, pageSize), cancellationToken);
            if (result is JsonElement json)
            {
                return json.Deserialize<PaginatedResponse<Brand>>() ?? throw new InvalidOperationException("Failed to deserialize response");
            }
            throw new InvalidOperationException("Unexpected response type");
        }
        /// <summary>
        /// Retrieves a list of item categories.
        /// </summary>
        /// <param name="licenseNumber">If specified, the categories will be retrieved for the specified License Number. If not specified, all item categories will be returned.</param>
        /// <param name="pageNumber">The number of the data page from which to return data.</param>
        /// <param name="pageSize">The number of records to return per page. Pagination is currently disabled by default. You can enable pagination on this query by specifying a value that does not exceed 20.</param>
        /// <param name="cancellationToken">The cancellation token to cancel the operation.</param>
        public async Task<PaginatedResponse<Category>> GetItemsCategories(string? licenseNumber = null, string? pageNumber = null, string? pageSize = null, CancellationToken cancellationToken = default)
        {
            var result = await _rateLimiter.ExecuteAsync<object>(null, true, () => _client.GetItemsCategories(licenseNumber, pageNumber, pageSize), cancellationToken);
            if (result is JsonElement json)
            {
                return json.Deserialize<PaginatedResponse<Category>>() ?? throw new InvalidOperationException("Failed to deserialize response");
            }
            throw new InvalidOperationException("Unexpected response type");
        }
        /// <summary>
        /// Retrieves a file by its Id for the specified Facility.
        /// Permissions Required:
        /// - Manage Items
        /// </summary>
        /// <param name="id">id</param>
        /// <param name="licenseNumber">The License Number of the Facility for which to return the file.</param>
        /// <param name="cancellationToken">The cancellation token to cancel the operation.</param>
        public async Task<File> GetItemsFileById(string id, string? licenseNumber = null, CancellationToken cancellationToken = default)
        {
            var result = await _rateLimiter.ExecuteAsync<object>(null, true, () => _client.GetItemsFileById(id, licenseNumber), cancellationToken);
            if (result is JsonElement json)
            {
                return json.Deserialize<File>() ?? throw new InvalidOperationException("Failed to deserialize response");
            }
            throw new InvalidOperationException("Unexpected response type");
        }
        /// <summary>
        /// Retrieves a list of inactive items for the specified Facility.
        /// Permissions Required:
        /// - Manage Items
        /// </summary>
        /// <param name="licenseNumber">The License Number of the Facility for which to return the list of inactive items.</param>
        /// <param name="pageNumber">The number of the data page from which to return data.</param>
        /// <param name="pageSize">The number of records to return per page. Pagination is currently disabled by default. You can enable pagination on this query by specifying a value that does not exceed 20.</param>
        /// <param name="cancellationToken">The cancellation token to cancel the operation.</param>
        public async Task<PaginatedResponse<Item>> GetInactiveItems(string? licenseNumber = null, string? pageNumber = null, string? pageSize = null, CancellationToken cancellationToken = default)
        {
            var result = await _rateLimiter.ExecuteAsync<object>(null, true, () => _client.GetInactiveItems(licenseNumber, pageNumber, pageSize), cancellationToken);
            if (result is JsonElement json)
            {
                return json.Deserialize<PaginatedResponse<Item>>() ?? throw new InvalidOperationException("Failed to deserialize response");
            }
            throw new InvalidOperationException("Unexpected response type");
        }
        /// <summary>
        /// Retrieves detailed information about a specific Item by Id.
        /// Permissions Required:
        /// - Manage Items
        /// </summary>
        /// <param name="id">id</param>
        /// <param name="licenseNumber">If specified, the Item will be validated against the specified License Number. If not specified, the Item will be validated against all of the User's current Facilities. Please note that if the Item is not valid for the specified License Number, a 401 Unauthorized status will be returned.</param>
        /// <param name="cancellationToken">The cancellation token to cancel the operation.</param>
        public async Task<Item> GetItemsById(string id, string? licenseNumber = null, CancellationToken cancellationToken = default)
        {
            var result = await _rateLimiter.ExecuteAsync<object>(null, true, () => _client.GetItemsById(id, licenseNumber), cancellationToken);
            if (result is JsonElement json)
            {
                return json.Deserialize<Item>() ?? throw new InvalidOperationException("Failed to deserialize response");
            }
            throw new InvalidOperationException("Unexpected response type");
        }
        /// <summary>
        /// Retrieves an image by its Id for the specified Facility.
        /// Permissions Required:
        /// - Manage Items
        /// </summary>
        /// <param name="id">id</param>
        /// <param name="licenseNumber">The License Number of the Facility for which to return the image.</param>
        /// <param name="cancellationToken">The cancellation token to cancel the operation.</param>
        public async Task<Photo> GetItemsPhotoById(string id, string? licenseNumber = null, CancellationToken cancellationToken = default)
        {
            var result = await _rateLimiter.ExecuteAsync<object>(null, true, () => _client.GetItemsPhotoById(id, licenseNumber), cancellationToken);
            if (result is JsonElement json)
            {
                return json.Deserialize<Photo>() ?? throw new InvalidOperationException("Failed to deserialize response");
            }
            throw new InvalidOperationException("Unexpected response type");
        }
        /// <summary>
        /// Updates one or more existing item brands for the specified Facility.
        /// Permissions Required:
        /// - Manage Items
        /// </summary>
        /// <param name="body">Request body.</param>
        /// <param name="licenseNumber">The License Number of the Facility for the Item Brand updates.</param>
        /// <param name="cancellationToken">The cancellation token to cancel the operation.</param>
        public async Task UpdateItemsBrand(List<ItemsUpdateBrandRequest>? body = null, string? licenseNumber = null, CancellationToken cancellationToken = default)
        {
            await _rateLimiter.ExecuteAsync<object?>(null, false, async () => { await _client.UpdateItemsBrand(body, licenseNumber); return null; }, cancellationToken);
        }
        /// <summary>
        /// Updates one or more existing products for the specified Facility.
        /// Permissions Required:
        /// - Manage Items
        /// </summary>
        /// <param name="body">Request body.</param>
        /// <param name="licenseNumber">The License Number of the Facility for the Item updates.</param>
        /// <param name="cancellationToken">The cancellation token to cancel the operation.</param>
        public async Task UpdateItems(List<ItemsUpdateItemsRequest>? body = null, string? licenseNumber = null, CancellationToken cancellationToken = default)
        {
            await _rateLimiter.ExecuteAsync<object?>(null, false, async () => { await _client.UpdateItems(body, licenseNumber); return null; }, cancellationToken);
        }
    }
}

